src/slurm_plugin/clustermgtd.py

```diff
@@ -0,0 +1,624 @@
+            return True
+        return False
+
+    def _handle_health_check(self, instance_health_states, id_to_instance_map, ip_to_slurm_node_map, health_check_type):
+        """
+        Perform health check action for all types of supported health checks.
+
+        Place nodes failing health check into DRAIN, so they can be maintained when possible
+        """
+        health_check_functions = {
+            ClusterManager.HealthCheckTypes.scheduled_event: {
+                "function": ClusterManager._fail_scheduled_events_check,
+                "default_kwargs": {},
+            },
+            ClusterManager.HealthCheckTypes.ec2_health: {
+                "function": ClusterManager._fail_ec2_health_check,
+                "default_kwargs": {
+                    "current_time": self.current_time,
+                    "health_check_timeout": self.sync_config.health_check_timeout,
+                },
+            },
+        }
+        log.info("Performing actions for health check type: %s", health_check_type)
+        nodes_failing_health_check = []
+        # Pick corresponding health check function
+        is_instance_unhealthy = health_check_functions.get(health_check_type).get("function")
+        default_kwargs = health_check_functions.get(health_check_type).get("default_kwargs")
+        for instance in instance_health_states:
+            if is_instance_unhealthy(instance_health_state=instance, **default_kwargs):
+                unhealthy_node = ip_to_slurm_node_map.get(id_to_instance_map.get(instance.id).private_ip)
+                if unhealthy_node:
+                    log.warning(
+                        "Node %s(%s) is associated with instance %s that is failing %s. EC2 health state: %s",
+                        unhealthy_node.name,
+                        unhealthy_node.nodeaddr,
+                        instance.id,
+                        health_check_type,
+                        instance,
+                    )
+                    nodes_failing_health_check.append(unhealthy_node.name)
+        if nodes_failing_health_check:
+            # Place unhealthy node into drain, this operation is idempotent
+            set_nodes_drain(nodes_failing_health_check, reason=f"Node failing {health_check_type}")
+
+    def _update_static_nodes_in_replacement(self, slurm_nodes):
+        """Update self.static_nodes_in_replacement by removing nodes that finished replacement and is up."""
+        nodename_to_slurm_nodes_map = {node.name: node for node in slurm_nodes}
+        nodes_not_in_replacement = set()
+        for nodename in self.static_nodes_in_replacement:
+            node = nodename_to_slurm_nodes_map.get(nodename)
+            # Remove nodename from static_nodes_in_replacement if node is no longer an active node or node is up
+            if not node or node.is_up():
+                nodes_not_in_replacement.add(nodename)
+        self.static_nodes_in_replacement.difference_update(nodes_not_in_replacement)
```

Comment:
minor: instead of creating a new list of nodes to remove and then compute the difference you could have computed the update list directly (by using the complementary condition `if node and not node.is_up()`) and then assign the new list.

Reply:
Good catch